llvm/include/llvm/IR/Intrinsics.td

```diff
@@ -700,10 +700,13 @@ class MSBuiltin<string name> {
 //===--------------- Variable Argument Handling Intrinsics ----------------===//
 //
 
-def int_vastart : DefaultAttrsIntrinsic<[], [llvm_ptr_ty], [], "llvm.va_start">;
-def int_vacopy  : DefaultAttrsIntrinsic<[], [llvm_ptr_ty, llvm_ptr_ty], [],
-                            "llvm.va_copy">;
-def int_vaend   : DefaultAttrsIntrinsic<[], [llvm_ptr_ty], [], "llvm.va_end">;
+def int_vastart : DefaultAttrsIntrinsic<[],
+                                        [llvm_anyptr_ty], [], "llvm.va_start">;
+def int_vacopy  : DefaultAttrsIntrinsic<[],
+                                        [llvm_anyptr_ty, llvm_anyptr_ty], [],
+                                        "llvm.va_copy">;
```

Comment:
vacopy probably shouldn't have 2 type parameters. 
```suggestion
def int_vacopy  : DefaultAttrsIntrinsic<[],
                                        [llvm_anyptr_ty, LLVMMatchType<0>], [],
                                        "llvm.va_copy">;
```

Reply:
Hmm, I did struggle a bit with this and admit that I'm not (yet) entirely convinced a valid (albeit hypothetical and admittedly odd) case couldn't be constructed. Consider e.g. having a `__builtin_va_list` declared at namespace / global scope. If a target uses an explicit, non-generic, AS, for globals, then the copy would be to/from a pointer to generic (or pointer to private) from/to a pointer to global. I _believe_ this should work, and making `va_copy` mono-parameter would break it, would it not?

Reply:
It does for our target, but I believe each target has freedom when it comes to implementing the actual `va_list` underlying type (the standard regards it as opaque), so it is probably legal to do things that do not match that. So I'd rather follow up with some AMDGPU specific work, but leave this as general / generic as possible.

Reply:
clang semantic analysis currently forces both the source and destination of va_copy to the type __builtin_va_list_ref, whatever that is.  So clang would never emit va_copy with different pointer types.  (Some other frontend could theoretically emit such operations, but it's hard for me to imagine it being useful.)

In theory, you could try to optimize the address-spaces involved in a va_copy operation... but there's not really much point to trying to optimize va_copy before you lower it.

Reply:
Ok, I think you and @arsenm have the right of it, I've taken Matt's suggestion.